python_modules/dagster/dagster/_core/definitions/decorators/asset_decorator.py

```diff
@@ -643,53 +666,79 @@ def inner(fn: Callable[..., Any]) -> AssetsDefinition:
+                node.asset_key: node.freshness_policy
+                for node in specs
+                if node.freshness_policy is not None
+            }
+            auto_materialize_policies_by_key = {
+                node.asset_key: node.auto_materialize_policy
+                for node in specs
+                if node.auto_materialize_policy is not None
+            }
+            metadata_by_key = {
+                node.asset_key: node.metadata for node in specs if node.metadata is not None
+            }
+            internal_deps = {node.asset_key: node.deps for node in specs if node.deps is not None}
+        else:
+            # source group names from the AssetOuts (if any)
+            group_names_by_key = {
+                keys_by_output_name[output_name]: out.group_name
+                for output_name, out in asset_out_map.items()
+                if out.group_name is not None
+            }
+            if group_name:
+                check.invariant(
+                    not group_names_by_key,
+                    "Cannot set group_name parameter on multi_asset if one or more of the AssetOuts"
+                    " supplied to this multi_asset have a group_name defined.",
+                )
+                group_names_by_key = {
+                    asset_key: group_name for asset_key in keys_by_output_name.values()
+                }
 
-        partition_mappings = {
-            keys_by_input_name[input_name]: asset_in.partition_mapping
-            for input_name, asset_in in (ins or {}).items()
-            if asset_in.partition_mapping is not None
-        }
-        metadata_by_key = {
-            keys_by_output_name[output_name]: out.metadata
-            for output_name, out in outs.items()
-            if out.metadata is not None
-        }
+            # source freshness policies from the AssetOuts (if any)
+            freshness_policies_by_key = {
+                keys_by_output_name[output_name]: out.freshness_policy
+                for output_name, out in asset_out_map.items()
+                if out.freshness_policy is not None
+            }
+            auto_materialize_policies_by_key = {
+                keys_by_output_name[output_name]: out.auto_materialize_policy
+                for output_name, out in asset_out_map.items()
+                if out.auto_materialize_policy is not None
+            }
+
+            partition_mappings = {
+                keys_by_input_name[input_name]: asset_in.partition_mapping
+                for input_name, asset_in in (ins or {}).items()
+                if asset_in.partition_mapping is not None
+            }
+            metadata_by_key = {
+                keys_by_output_name[output_name]: out.metadata
+                for output_name, out in asset_out_map.items()
+                if out.metadata is not None
+            }
+            internal_deps = {keys_by_output_name[name]: asset_deps[name] for name in asset_deps}
```

Comment:
this might be enough parallel values to warrant a temporary dataclass/namedtuple